Adding hooks introduction documentation (#18)
babbel/build-your-own-react-workshop.js: docs/hooks.md

# Hooks

> Hooks are the current way to make components a little more useful, they generally allow functional components to have side effects such as keeping state or running side effects after rendering.

Reviewer: Can remove the double use of side effects here 🤪
```suggestion
Hooks have made functional components a lot more useful in React apps, allowing them to be stateful  and run side effects after rendering or on prop changes.
```

> For the sake of time and not being too overtly ambitious, within our own React we decided to focus on two of the most used hooks:

Reviewer: ```suggestion
For the sake of time, we decided to focus on two of the most used hooks within our own React:
```

Reviewer: We are ambitious! Don't let them think otherwise!

Author: Haha very true, given this is somewhat hard to get through, not saying that is an amazing idea ❤️

> - `useState` to allow components to hold state that can be updated and will trigger updates

Reviewer: ```suggestion
- `useState` to allow components to hold state that can be updated and can trigger further updates.
```

- `useEffect` to run side effects after render phases

## Anatomy of a hook

Before starting to think about implementation, let's first think a bit about how any hook normally works.

> We will take `useState` to make the example a bit more concrete, but for the portion of this chapter none of the elements will be specific to `useState`.

Reviewer: Maybe something more like "We will take `useState` as an example but the elements and logic described applies for all hooks"

```javascript
// We always import a react hook as a named export from react
import { useState } from 'react';

// Using a hook outside of a component is forbidden
useState(test); // THIS IS ILLEGAL

const Counter = () => {
    // we must use the hook in a react component
    const [count, setCount] = useState(0);
    // hooks have no identification, react relies on order of calling
    const [anotherState, setAnotherState] = useState(0);
    if (anotherState > 0) {
        // which explains the rule of hooks that hooks can't be assign conditionally
        const [yetAnotherState, setYetAnotherState] = useState(0);
    }

    return (
        <div>
            {count}
            <button onClick={() => setCount(count => count + 1)}>+</button>
        </div>
    );
};
```

> So the minimal pieces of a hook anatomy are the following:

Reviewer: ```suggestion
The fundamentals of a hook are:
```

1. A react hook is exported as a named export by React

> 2. It must be called in a React component

Reviewer: functional component?
